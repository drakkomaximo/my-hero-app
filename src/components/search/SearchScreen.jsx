import React, { useMemo } from "react";
import queryString from "query-string";
import { useLocation } from "react-router";
import { useCustomForm } from "../../hooks/useCustomForm";
import HeroCard from "../heroes/HeroCard";
import getHeroesByName from "../selectors/getHeroesByName";

const SearchScreen = ({ history }) => {
  const location = useLocation();
  const { q = "" } = queryString.parse(location.search);

  const [formValues, handleInputChange, reset] = useCustomForm({
    searchHero: '',
  });

  const { searchHero } = formValues;

  const heroesFilted = useMemo(() => getHeroesByName(q), [q]);

  const HandleSearch = (e) => {
    e.preventDefault();
    history.push(`?q=${searchHero}`);
    console.log(searchHero)
    reset();
  };

  return (
    <div>
      <h1>Search Screen</h1>
      <hr />

      <div className="row animate__animated animate__fadeIn">
        <div className="col-5">
          <h4>Search Form</h4>
          <hr />
          <form onSubmit={HandleSearch}>
            <input
              autoComplete="off"
              type="text"
              placeholder="Find your hero"
              className="form-control"
              name="searchHero"
              onChange={handleInputChange}
              value={searchHero}
            />
            <button
              type="submit"
              className="btn m-1 btn-block btn-outline-primary"
            >
              Search...
            </button>
          </form>
        </div>
        <div className="col-7">
          <h4> Results </h4>
          <hr />

          {q === "" && <div className="alert alert-info">Search a hero</div>}
          {q !== "" && heroesFilted.length === 0 && <div className="alert alert-danger">There is no a hero with {q}</div>}
          {heroesFilted.map((hero) => (
            <HeroCard key={hero.id} {...hero} />
          ))}
        </div>
      </div>
    </div>
  );
};

export default SearchScreen;
