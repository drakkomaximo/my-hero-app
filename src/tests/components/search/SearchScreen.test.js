import { mount } from "enzyme"
import { MemoryRouter, Route } from "react-router"
import { AuthContext } from "../../../auth/AuthContex"
import SearchScreen from "../../../components/search/SearchScreen"

describe('Tests on <SearchScreen />', () => {

    const historyMock = {
        push: jest.fn()
    }

    const wrapper = mount(
        <MemoryRouter initialEntries={['/search']}>
            <Route path='/search' component={SearchScreen} />
        </MemoryRouter>
    )

    test('It must show the default values correctly', () => {
        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find('.alert').text().trim()).toBe('Search a hero')
    })

    test('It must activate reset function', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path='/search' component={SearchScreen} />
            </MemoryRouter>
        )

        expect( wrapper.find('input').prop('value') ).toBe('')
        expect( wrapper.find('HeroCard').exists() ).toBe(true)
    })

    test('It must show an error if a hero does not exist', () => {
        const queryString = 'batmanwerewrwerwe'
        const wrapper = mount(
            <MemoryRouter initialEntries={[`/search?q=${queryString}`]}>
                <Route path='/search' component={SearchScreen} />
            </MemoryRouter>
        )
        expect(wrapper.find('.alert-danger').text().trim()).toBe(`There is no a hero with ${queryString}`)
    })

    test('It must call push function of history', () => {
        const queryString = 'batman'
        const wrapper = mount(
            <MemoryRouter initialEntries={[`/search?q=${queryString}`]}>
                <Route path='/search' component={() => <SearchScreen history={ historyMock }/>} />
            </MemoryRouter>
        )

        wrapper.find('input').simulate('change', {
            target: {
                name:'searchHero',
                value: queryString
            }
        })

        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        })

        expect( historyMock.push ).toHaveBeenCalledWith(`?q=${queryString}`)
    })
    
    


})
