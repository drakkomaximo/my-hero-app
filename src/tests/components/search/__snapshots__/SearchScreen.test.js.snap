// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <SearchScreen /> It must show the default values correctly 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row animate__animated animate__fadeIn"
  >
    <div
      className="col-5"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchHero"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-outline-primary"
          type="submit"
        >
          Search...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
         Results 
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero
      </div>
    </div>
  </div>
</div>
`;
